broker-auth-local/src/main/java/org/aktin/broker/auth/apikey/ApiKeyPropertiesAuthProvider.java

```diff
@@ -2,36 +2,180 @@
-//			System.out.println("Rewritten "+count+" node DN strings.");
-//		}
-	}
+
+  private PropertyFileAPIKeys keys;
+
+  /**
+   * Constructs a new ApiKeyPropertiesAuthProvider with lazy initialization. The API keys will be loaded later when getInstance() is called.
+   */
+  public ApiKeyPropertiesAuthProvider() {
+    keys = null;
+  }
+
+  /**
+   * Constructs a new ApiKeyPropertiesAuthProvider and immediately loads API keys from the given input stream.
+   *
+   * @param in The input stream containing the API keys properties.
+   * @throws IOException If an I/O error occurs while reading from the input stream.
+   */
+  public ApiKeyPropertiesAuthProvider(InputStream in) throws IOException {
+    this.keys = new PropertyFileAPIKeys(in);
+  }
+
+  /**
+   * Retrieves or initializes the PropertyFileAPIKeys instance.
+   *
+   * @return The PropertyFileAPIKeys instance.
+   * @throws IOException If an I/O error occurs while loading the properties file.
+   */
+  @Override
+  public PropertyFileAPIKeys getInstance() throws IOException {
+    if (this.keys == null) {
+      try (InputStream in = Files.newInputStream(getPropertiesPath())) {
+        this.keys = new PropertyFileAPIKeys(in);
+      }
+    }
+    return keys;
+  }
+
+  /**
+   * @return The Path object representing the location of API keys properties file
+   */
+  private Path getPropertiesPath() {
+    return path.resolve("api-keys.properties");
+  }
+
+  /**
+   * Adds a new API key and updates the properties file.
+   *
+   * @param apiKey   The new API key to add.
+   * @param clientDn The distinguished name of the client associated with the API key.
+   * @throws IOException              If an I/O error occurs while updating the properties file.
+   * @throws IllegalArgumentException If the API key already exists.
+   * @throws IllegalStateException    If the PropertyFileAPIKeys instance is not initialized.
+   */
+  public void addNewApiKeyAndUpdatePropertiesFile(String apiKey, String clientDn) throws IOException {
+    checkKeysInitialized();
+    Properties properties = keys.getProperties();
+    if (properties.containsKey(apiKey)) {
+      throw new IllegalArgumentException("API key already exists: " + apiKey);
+    }
+    keys.putApiKey(apiKey, clientDn);
+    saveProperties(keys);
+  }
+
+  /**
+   * Sets the state of an existing API key and updates the properties file.
+   *
+   * @param apiKey The API key to update.
+   * @param status The new status to set for the API key.
+   * @throws IOException            If an I/O error occurs while updating the properties file.
+   * @throws NoSuchElementException If the API key does not exist.
+   * @throws SecurityException      If an attempt is made to modify an admin API key.
+   * @throws IllegalStateException  If the PropertyFileAPIKeys instance is not initialized.
+   */
+  public void setStateOfApiKeyAndUpdatePropertiesFile(String apiKey, ApiKeyStatus status) throws IOException {
```

Comment:
You should add the other exceptions also to the throws clause in the function definition (not only to the documentation)